substratevm/src/com.oracle.svm.agent.core/src/com/oracle/svm/agent/core/JvmtiAgentBase.java

```diff
@@ -0,0 +1,275 @@
+ * (parameters and the return type) of the desired JVMTI event. The method must also be annotated
+ * with {@link CEntryPointOptions} and must use the
+ * {@link com.oracle.svm.agent.core.AgentIsolate.Prologue} prologue.</li>
+ * <li>Create a {@link CEntryPointLiteral} from your callback method.</li>
+ * <li>Set the callback in the callbacks argument of your {@link #onLoadCallback} method.</li>
+ * <lI>Enable the event using {@link com.oracle.svm.agent.core.jvmti.JvmtiEventMode}.</lI>
+ * </ol>
+ *
+ * Only one agent can be created in a single native image.
+ *
+ * @param <T> The subclass of {@link JNIHandleSetBase} containg the JNI handles this agent requires.
+ */
+@SuppressWarnings("JavadocReference")
+public abstract class JvmtiAgentBase<T extends JNIHandleSetBase> {
+
+    private static final CEntryPointLiteral<CFunctionPointer> onVMInitLiteral = CEntryPointLiteral.create(JvmtiAgentBase.class, "onVMInit", JvmtiEnv.class, JNIEnvironment.class,
+                    JNIObjectHandle.class);
+    private static final CEntryPointLiteral<CFunctionPointer> onVMStartLiteral = CEntryPointLiteral.create(JvmtiAgentBase.class, "onVMStart", JvmtiEnv.class, JNIEnvironment.class);
+    private static final CEntryPointLiteral<CFunctionPointer> onVMDeathLiteral = CEntryPointLiteral.create(JvmtiAgentBase.class, "onVMDeath", JvmtiEnv.class, JNIEnvironment.class);
+    private static final CEntryPointLiteral<CFunctionPointer> onThreadEndLiteral = CEntryPointLiteral.create(JvmtiAgentBase.class, "onThreadEnd", JvmtiEnv.class, JNIEnvironment.class,
+                    JNIObjectHandle.class);
+
+    @SuppressWarnings({"rawtypes"}) private static JvmtiAgentBase singleton;
+
+    /**
+     * Callback method that should create your subclass of {@link JNIHandleSetBase}.
+     *
+     * @param env JNI environment of the thread running the JVMTI callback.
+     * @return An instance of your {@link JNIHandleSetBase} subclass.
+     */
+    protected abstract JNIHandleSetBase constructJavaHandles(JNIEnvironment env);
+
+    /**
+     * JVMTI Agent_OnLoad callback.
+     *
+     * @param vm The JNI Java VM.
+     * @param jvmti The JVMTI environment.
+     * @param callbacks A structure that allows setting the callbacks for JVMTI events.
+     * @param options Command line options passed to the agent.
+     * @return 0 on success, anything else on failure.
+     */
+    protected abstract int onLoadCallback(JNIJavaVM vm, JvmtiEnv jvmti, JvmtiEventCallbacks callbacks, String options);
+
+    /**
+     * JVMTI Agent_OnUnload callback.
+     *
+     * @param vm The JNI Java VM.
+     * @return 0 on success, anything else on failure.
+     */
+    protected abstract int onUnloadCallback(JNIJavaVM vm);
+
+    /**
+     * JVMTI VMStart event callback.
+     *
+     * @param jvmti The JVMTI environment.
+     * @param jni The JNI environment of the thread running the JVMTI callback.
+     */
+    protected abstract void onVMStartCallback(JvmtiEnv jvmti, JNIEnvironment jni);
```

Comment:
In this API I find inconsistencies with capitalization: why is VM here, and Vm later. Why is it `JNIJavaVM` but then we have a `JvmtiEnv`. We should stick to one naming convention in this API, whichever it is. \CC @peter-hofer

Reply:
`JNIEnvironment`, `JNIJavaVM` etc. are classes from our JNI implementation. JNI and JVMTI identifiers themselves also differ in that aspect: `jvmtiEnv` vs `JNIEnv`.

Reply:
Well, if it is `onVMStartCallback` then it should be `onVMDeathCallback` instead of `onVmDeathCallback`. I will change this when merging.

Reply:
No argument there.

Reply:
I have updated it to `onVMDeathCallback`, that was an oversight on my part. I have also fixed some javadoc related warnings for `@link` embedded inside a `@see`.